From 278dbbb39d05cccc6b9131371dce8edaddf9f7ef Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Wed, 24 Sep 2014 14:05:30 +0800
Subject: [PATCH 0114/5242] MLK-9606-2 usb: otg: test: start tst_maint timer
 and set otg_vbus_off flag

commit  3151e88af1434c97f257c95ec6a470aa71c90b18 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds 2 variables: tst_maint and otg_vbus_off, tst_maint is to check if
current session for test device; otg_vbus_off is to notify if A device need turn
off vbus immediately after B device disconnects.
The otg test device handling is added into ehset driver, for that device,
A-device should start a timer for maintain the session, and set otg_vbus_off
flag according to its bcdDevice value.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 78a2c14bdea4b8e334c7e0afad074b61b71193cb)
(cherry picked from commit b854141cc5a7150751e7f7fd2dfb10a415cfea10)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/misc/ehset.c    |   18 ++++++++++++++++++
 include/linux/usb/otg-fsm.h |    5 +++++
 2 files changed, 23 insertions(+)

diff --git a/drivers/usb/misc/ehset.c b/drivers/usb/misc/ehset.c
index 7895d61..ffea8015 100644
--- a/drivers/usb/misc/ehset.c
+++ b/drivers/usb/misc/ehset.c
@@ -9,6 +9,7 @@
 #include <linux/slab.h>
 #include <linux/usb.h>
 #include <linux/usb/ch11.h>
+#include <linux/usb/otg-fsm.h>
 
 #define TEST_SE0_NAK_PID			0x0101
 #define TEST_J_PID				0x0102
@@ -17,6 +18,7 @@
 #define TEST_HS_HOST_PORT_SUSPEND_RESUME	0x0106
 #define TEST_SINGLE_STEP_GET_DEV_DESC		0x0107
 #define TEST_SINGLE_STEP_SET_FEATURE		0x0108
+#define TEST_OTG_TEST_DEVICE_SUPPORT		0x0200
 
 static int ehset_probe(struct usb_interface *intf,
 		       const struct usb_device_id *id)
@@ -27,6 +29,7 @@ static int ehset_probe(struct usb_interface *intf,
 	struct usb_device_descriptor *buf;
 	u8 portnum = dev->portnum;
 	u16 test_pid = le16_to_cpu(dev->descriptor.idProduct);
+	struct otg_fsm *fsm = dev->bus->otg_fsm;
 
 	switch (test_pid) {
 	case TEST_SE0_NAK_PID:
@@ -107,6 +110,20 @@ static int ehset_probe(struct usb_interface *intf,
 					NULL, 0, 60 * 1000);
 
 		break;
+	case TEST_OTG_TEST_DEVICE_SUPPORT:
+		if (!fsm || dev->bus->is_b_host)
+			return ret;
+
+		mutex_lock(&fsm->lock);
+		fsm->tst_maint = 1;
+		if (le16_to_cpu(dev->descriptor.bcdDevice) & 0x1)
+			fsm->otg_vbus_off = 1;
+		else
+			fsm->otg_vbus_off = 0;
+		mutex_unlock(&fsm->lock);
+		otg_add_timer(fsm, A_TST_MAINT);
+		ret = 0;
+		break;
 	default:
 		dev_err(&intf->dev, "%s: unsupported PID: 0x%x\n",
 			__func__, test_pid);
@@ -127,6 +144,7 @@ static void ehset_disconnect(struct usb_interface *intf)
 	{ USB_DEVICE(0x1a0a, TEST_HS_HOST_PORT_SUSPEND_RESUME) },
 	{ USB_DEVICE(0x1a0a, TEST_SINGLE_STEP_GET_DEV_DESC) },
 	{ USB_DEVICE(0x1a0a, TEST_SINGLE_STEP_SET_FEATURE) },
+	{ USB_DEVICE(0x1a0a, TEST_OTG_TEST_DEVICE_SUPPORT) },
 	{ }			/* Terminating entry */
 };
 MODULE_DEVICE_TABLE(usb, ehset_id_table);
diff --git a/include/linux/usb/otg-fsm.h b/include/linux/usb/otg-fsm.h
index 0d318f7..0c8b3cc 100644
--- a/include/linux/usb/otg-fsm.h
+++ b/include/linux/usb/otg-fsm.h
@@ -55,6 +55,7 @@ enum otg_fsm_timer {
 	B_DATA_PLS,
 	B_SSEND_SRP,
 	A_DP_END,
+	A_TST_MAINT,
 
 	NUM_OTG_FSM_TIMERS,
 };
@@ -173,6 +174,10 @@ struct otg_fsm {
 	int b_hnp_enable;
 	int a_clr_err;
 
+	/* OTG test device */
+	int tst_maint;
+	int otg_vbus_off;
+
 	/* Informative variables. All unused as of now */
 	int a_bus_drop_inf;
 	int a_bus_req_inf;
-- 
1.7.9.5

